### PR TITLE
add section on LLM-as-rater

### LINKED ISSUE #27: Update LLM as Rater section
- [x] Fix the todos in Judges/Raters
- [x] Add examples
- [x] Flesh out advantages and challenges

### PR DESCRIPTION
**Pull request by neilernst**
fixes #27

### CODE REVIEW

**Comment by neilernst** [comment]
@ctreude you have a ~~recent~~ forthcoming MSR paper on this exact topic ...  would appreciate your feedback.

**Comment by neilernst** [comment]
I also have not previewed this with Jekyll - formatting e.g. for the prompt example may be wrong.

File: study-types/_sources/01_study-types.tex [hunk 1]
```diff
@@ -20,7 +20,7 @@ \subsection{Introduction: LLMs as Tools for Software Engineering Researchers}
 
 
 \subsection{LLMs as Annotators}
-
+\label{sec:annotate}
```
Comment (sbaltes) [inline study-types/_sources/01_study-types.tex:23]: I'm not sure whether this works on the website. You might have to use paths relative to the website root: https://github.com/se-ubt/llm-guidelines-website/blob/ed4d9249a79df2596a80873d74a74369f0cb965f/guidelines/_sources/01_declare-usage-and-role.tex#L43

File: literature.bib [hunk 2]
```diff
@@ -1124,3 +1124,132 @@ @article{widder2024open
...
+
+@article{Gallegos2023BiasAF,
+  title={Bias and Fairness in Large Language Models: A Survey},
+  author={Isabel O. Gallegos and Ryan A. Rossi and Joe Barrow and Md. Mehrab Tanjim and Sungchul Kim and Franck Dernoncourt and Tong Yu and Ruiyi Zhang and Nesreen Ahmed},
+  journal={Computational Linguistics},
+  year={2023},
+  volume={50},
+  pages={1097-1179},
+  url={https://api.semanticscholar.org/CorpusID:261530629}
+}
```
Comment (sbaltes) [inline literature.bib:1255]: I think we agreed to use DBLP BibTex if available. Maybe we have to make that more explicit somewhere. Not a major issue, but if you find the time, it would be great to update the references accordingly.

File: study-types/_sources/01_study-types.tex [hunk 3]
```diff
@@ -58,24 +58,64 @@ \subsubsection{Challenges}
...
+
+% Other risks or issues
+% - Correlation scores like K-Alpha may assume human raters, not machines https://arxiv.org/abs/2410.03775
+% - how much context to provide, how to use RAG
 
 \subsubsection{Advantages}
 
 By providing---depending on the model configuration---consistent and relatively ``objective'' evaluations, LLMs can help mitigate certain biases and part of the variability that human raters might introduce. 
-This can lead to more reliable and reproducible results in empirical studies.
+This may lead to more reliable and reproducible results in empirical studies, to the extent these models can be reproduced or checkpointed. % TODO: add reference to relevant guideline?
```
Comment (sbaltes) [inline study-types/_sources/01_study-types.tex:104]: I think this is something we need to agree on in general. My current thinking was to keep the study types independent and only reference the study types from the guidelines, not the other way around. But something we can discuss in the next meeting.

**Review by sbaltes — COMMENTED** [review]
Added some comments, @ctreude can have a closer look at the content.

**Comment by sbaltes** [comment]
Here's the resulting Markdown:

## LLMs as Raters

### Description

In empirical studies, LLMs can act as raters (also known as judges) to
evaluate properties of software artifacts. For instance, LLMs can be
used to assess code readability, adherence to coding standards, or the
quality of code comments. *Rating* is distinct from the more qualitative
task of assigning a code or label (see
<a href="#sec:annotate" data-reference-type="ref"
data-reference="sec:annotate">0.2</a>) to (typically) unstructured text.
It is also distinct from the software engineering tasks of **using**
LLMs, as we discuss in <a href="#sec:llm4se" data-reference-type="ref"
data-reference="sec:llm4se">[sec:llm4se]</a> to e.g. understand if LLMs
can find ambiguity in requirements \[12\], which is then compared to a
human-created gold standard.

### Example(s)

Lubos et al. \[13\] leveraged Llama-2[1] to evaluate the quality of
software requirements statements.

They prompted the LLM with the text below, where the words in brackets
reflect parameters for the study:

> Your task is to evaluate the quality of a software requirement.  
> Evaluate whether the following requirement is
> {quality\_characteristic}. {quality\_ characteristic} means:
> {quality\_characteristic\_explanation}  
> The evaluation result must be: ‘yes’ or ‘no’.  
> Request: Based on the following description of the project:
> {project\_description} Evaluate the quality of the following
> requirement: {requirement}.  
> Explain your decision and suggest an improved version.  

Lubos et al. then evaluated the LLM’s output against human raters, to
assess how well the LLM matched experts. They did this in 2 phases: in
one, the humans were unaware of the LLM’s decision, and in the other,
with foreknowledge. This second aspect measured the extent to which the
LLM might *convince* the human to change their mind.

Agreement can be measured in many ways; this study used the conventional
Cohen’s kappa measure \[14\]. The study found moderate agreement on the
simple requirements dataset, and poor agreement on the more complex
requirements. However, evaluating machine ratings may not be the same as
human-human ratings and is an open area of research. See Elangovan et
al. \[15\] for more.

Another study by Ahmed et al. \[16\] looked at LLMs-as-raters across
five datasets and ten annotation tasks. They used a variety of different
LLMs to compare model to model performance. For tasks where humans do
not agree with one another (in this case, evaluating conciseness of code
summaries) models also do poorly; but “if multiple LLMs reach similar
solutions independently, then LLMs are likely suitable for the
annotation task \[16, p. 6\].

### Advantages

By providing—depending on the model configuration—consistent and
relatively “objective” evaluations, LLMs can help mitigate certain
biases and part of the variability that human raters might introduce.
This may lead to more reliable and reproducible results in empirical
studies, to the extent these models can be reproduced or checkpointed.

LLMs can be much more efficient than the equivalent human approach. The
main constraint, which varies by model and budget, is the input context
size, i.e., the number of tokens one can pass into a model. For example,
the upper bound on context that can be passed into OpenAi’s
<span class="sans-serif">o1-mini</span> model is 32k[2].

### Challenges

When relying on the judgment of LLMs, researchers must build a reliable
process for generating ratings that considers the non-deterministic
nature of LLMs and report the intricacies of that process
transparently \[17\].

Avoid the temptation to assign objectivity to an LLM; evidence shows
LLMs can behave differently, for instance, when reviewing their own
outputs \[18\]. In more human-oriented datasets (such as discussions of
pull requests) LLMs may suffer from well documented biases and issues
with fairness \[19\].

Rating large numbers of items—for example, to perform fault localization
on the thousands of bugs big open-source projects deal with—comes with
costs in clock time, compute time, and environmental sustainability.

One researcher choice in such studies is the number of examples
(training data) provided to the LLM. This might involve zero tuning
(zero-shot), several example (few-shot), or more conventional training
data with many examples. In the example above, Lubos et al. chose
zero-shot tuning, providing no specific guidance besides the project’s
context, i.e., they did not show the LLM what a yes or no answer might
look like.

Research is ongoing as to how suitable LLMs are as standalone raters.
Questions around bias, accuracy, and trust remain \[20\]. There is
reason for concern about LLMs judging student assignments or doing peer
review of scientific papers \[21\]. Even beyond questions of technical
capacity, ethical questions remain, particularly if there is some
implicit expectation that a human is rating the output, such as a quiz.
Involving a human in the rating loop—for example, to contextualize the
scoring—is one approach \[22\].

File: study-types/_sources/01_study-types.tex (hunk 3 above)
Comment (neilernst) [reply study-types/_sources/01_study-types.tex:104]: maybe it is enough for guidelines to refer to example study types (one way relationship)

Comment (sbaltes) [reply study-types/_sources/01_study-types.tex:104]: Yeah, that's what I thought. But let's ask the others in the next meeting.

File: study-types/_sources/01_study-types.tex (hunk 1 above)
Comment (neilernst) [reply study-types/_sources/01_study-types.tex:23]: there are ways to do it - see https://github.com/fncokg/pandoc-tex-numbering - but maybe that is too much of a headache.

File: literature.bib (hunk 2 above)
Comment (sbaltes) [reply literature.bib:1255]: No worries. We need to make this more explicit and mention it again during the next meeting.

For example, the ahmed2025llmsreplacemanualannotation paper is already part of the library as DBLP:journals/corr/abs-2408-05534. So it's also about having unique cite keys that allow us to detect duplicates (with the same cite key Latex would have complained during the compilation).

Here's another example from the papers you added: https://dblp.org/rec/conf/coling/ZhouC024.html?view=bibtex
Copying and pasting this data from DBLP instead of anywhere else ensures using the same abbreviations for conference and journals names, unique cite keys, etc.

Does this make sense to you?

File: study-types/_sources/01_study-types.tex (hunk 1 above)
Comment (sbaltes) [reply study-types/_sources/01_study-types.tex:23]: AFAIK this is not supported by Jekyll/Kramdown. As long as we don't have a good reason - and references are relatively rare - I personally would not change our current pipeline.

Comment (neilernst) [reply study-types/_sources/01_study-types.tex:23]: added the href

File: literature.bib [hunk 4]
```diff
@@ -1124,3 +1124,132 @@ @article{widder2024open
...
+ volume={20}, 
+ ISSN={1552-3888}, 
+ DOI={10.1177/001316446002000104}, number={1}, 
+ journal={Educational and Psychological Measurement}, 
+ publisher={SAGE Publications}, 
+ author={Cohen, Jacob}, 
+ year={1960}, month=apr, pages={37-46} }
+
+@article{Gallegos2023BiasAF,
+  title={Bias and Fairness in Large Language Models: A Survey},
```
Comment (neilernst) [inline literature.bib:1248]: not in dblp yet

File: literature.bib [hunk 5]
```diff
@@ -1124,3 +1124,132 @@ @article{widder2024open
...
+      author={Aparna Elangovan and Lei Xu and Jongwoo Ko and Mahsa Elyasi and Ling Liu and Sravan Bodapati and Dan Roth},
+      year={2025},
+      eprint={2410.03775},
+      archivePrefix={arXiv},
+      primaryClass={cs.HC},
+      url={https://arxiv.org/abs/2410.03775}, 
+}
+
+
+ @article{cohen60,
```
Comment (neilernst) [inline literature.bib:1237]: not in dblp

File: literature.bib [hunk 6]
```diff
@@ -1124,3 +1124,132 @@ @article{widder2024open
...
+}
+
+
+@inproceedings{NEURIPS2024_7f1f0218,
```
Comment (neilernst) [inline literature.bib:1183]: not yet in dblp as neurips paper

File: literature.bib [hunk 7]
```diff
@@ -1124,3 +1124,132 @@ @article{widder2024open
...
 }
 
+ @inproceedings{10.1145/3510003.3510157, 
+ title={Automated handling of anaphoric ambiguity in requirements: a multi-solution study}, 
+ url={http://dx.doi.org/10.1145/3510003.3510157}, DOI={10.1145/3510003.3510157}, 
+ booktitle={Proceedings of the 44th International Conference on Software Engineering}, publisher={ACM}, 
+ author={Ezzini, Saad and Abualhaija, Sallam and Arora, Chetan and Sabetzadeh, Mehrdad}, 
+ year={2022}, month=may, pages={187-199} }
+
+@inproceedings{panHumanCenteredDesignRecommendations2024,
```
Comment (neilernst) [inline literature.bib:1134]: not in dblp as workshop, just arxiv

File: literature.bib (hunk 4 above)
Comment (sbaltes) [reply literature.bib:1248]: That's fine. For everything that's not in DBLP, we just use whatever is available. Maybe search in the Bibtex file first whether someone else has already added the reference.

File: literature.bib (hunk 2 above)
Comment (sbaltes) [reply literature.bib:1255]: That's fine. For everything that's not in DBLP, we just use whatever is available. Maybe search in the Bibtex file first whether someone else has already added a certain reference to avoid duplicates.

**Comment by neilernst** [comment]
made the requested changes by @sbaltes - compiled and verified locally with Jekyll

**Comment by sbaltes** [comment]
Great, thanks! It seems you need to rebase before we can merge.

**Comment by neilernst** [comment]
I did earlier - says no conflicts with main for me. Am I missing something?

**Comment by sbaltes** [comment]
Not sure what the problem is, but GitHub says it can't rebase automatically: 
image

**Comment by neilernst** [comment]
ok, after some sort of recursive rebase hell I think it should work now ... not sure what I did.
